Add sshchroot connection plugin

aarnaud opened the pull request:
##### SUMMARY
Add sshchroot connection plugin

```
- name: Chroot in /mnt on server01
  host: server01:/mnt
  roles:
    - ....
```

##### COMPONENT NAME
sshchroot

aarnaud committed:
Add sshchroot connection plugin

Signed-off-by: Anthony ARNAUD <github@anthony-arnaud.fr>

maxamillion commented on lib/ansible/plugins/connection/sshchroot.py:
```diff
@@ -0,0 +1,227 @@
+    short_description: connect via ssh client binary
+    description:
+        - This connection plugin allows ansible to communicate to the target machines via normal ssh command line.
+    author: ansible (@core)
```
Update this to yourself. :)

maxamillion commented on lib/ansible/plugins/connection/sshchroot.py:
```diff
@@ -0,0 +1,227 @@
+    description:
+        - This connection plugin allows ansible to communicate to the target machines via normal ssh command line.
+    author: ansible (@core)
+    version_added: historical
```
This should reflect `2.8` as we're currently in the [Ansible 2.8 development cycle](https://docs.ansible.com/ansible/devel/roadmap/).

maxamillion commented on lib/ansible/plugins/connection/sshchroot.py:
```diff
@@ -0,0 +1,227 @@
+'''
+
+EXAMPLES = '''
+- name: Chroot in /mnt on server01
```
This should be a more comprehensive example, I don't follow what the intent here is. I have an assumption that you're connecting to a ssh daemon that's running in the context of a chroot located at `/mnt` on the `server01` system but it should be more explicitly stated in an attempt to remove the ambiguity.

maxamillion commented on lib/ansible/plugins/connection/sshchroot.py:
```diff
@@ -0,0 +1,227 @@
+class Connection(SSHConnection):
+    ''' SSH chroot based connections '''
+
+    chroot_bin = "/usr/sbin/chroot"
```
This should be configurable with a sane default (which I think `/usr/bin/chroot` is a sane default).

maxamillion reviewed (changes requested):
Thank you for the pull request! We'll need some [integration tests](https://docs.ansible.com/ansible/latest/dev_guide/testing_integration.html) and a [changelog fragment](https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html#changelogs).

mattclay reviewed (changes requested):
@aarnaud Can you add an integration test target for this connection plugin?

You'll want to start with the existing SSH test here:

https://github.com/ansible/ansible/tree/devel/test/integration/targets/connection_ssh/

Make a copy of that named after the connection plugin and update the inventory file, probably something like this:

```ini
[sshchroot]
sshchroot-pipelining    ansible_ssh_pipelining=true
sshchroot-no-pipelining ansible_ssh_pipelining=false
[sshchroot:vars]
ansible_host=localhost
ansible_connection=sshchroot
chroot_path=/
chroot_executable=/usr/sbin/chroot
```

That should at least get the test running on some platforms. You can test locally with:

`bin/ansible-test integration connection_sshchroot -v`

If you have Docker available I recommend trying with at least one of our test containers:

`bin/ansible-test integration connection_sshchroot -v --docker centos7`

The example above uses `/` for the chroot to avoid having to set up an actual chroot environment, as we do with the `chroot` connection plugin.

aarnaud commented:
@mattclay done and All check have passed !

mattclay commented on test/integration/targets/connection_sshchroot/aliases:
```diff
@@ -0,0 +1,3 @@
+needs/ssh
+needs/sshchroot
```
This line isn't needed, as it won't do anything. The `needs/ssh` line above it is a special alias which tells `ansible-test` it needs to make sure the local SSH service is responding before running the tests.

mattclay commented on test/integration/targets/connection_sshchroot/posix.sh:
```diff
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
```
Use a symbolic link to `../connection_posix/test.sh` as the other connection tests do, instead of copying the file.

aarnaud replied:
My bad, fixed

mattclay commented on test/integration/targets/connection_sshchroot/test_connection.inventory:
```diff
@@ -0,0 +1,7 @@
+ansible_host=localhost
+ansible_connection=sshchroot
+chroot_path=/
+chroot_executable=/usr/sbin/chroot
```
Include a single newline at the end of the file.

mattclay commented on test/integration/targets/connection_sshchroot/test_connection.inventory:
```diff
@@ -0,0 +1,7 @@
+[ssh]
+ssh-pipelining    ansible_ssh_pipelining=true
+[ssh:vars]
```
Use this instead:

```ini
[sshchroot]
sshchroot-pipelining    ansible_ssh_pipelining=true
sshchroot-no-pipelining ansible_ssh_pipelining=false
[sshchroot:vars]
```

This changes the labels from `ssh` to `sshchroot` and includes both the pipelining and no pipelining tests.

aarnaud replied:
Renamed, but i think i'm not support `no-pipelining`,

mattclay commented:
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/102368/66/tests

aarnaud commented:
Yes i will fix that in 2 weeks or maybe before

Akasurde commented:
@aarnaud Are you still working on this, let us know ? Thanks, needs_info

aarnaud commented:
Hi, i forgot to do update. 
It's seem the inventory host is not good because It try to use localhost connection instead of sshchroot and maybe localhost doesn't have ssh

aarnaud commented:
@mattclay I rebased my branch and fixed errors.
But for 2 tests on opensuse chroot is missing in the images.
https://app.shippable.com/github/ansible/ansible/runs/148299/100/console
https://app.shippable.com/github/ansible/ansible/runs/148299/99/console

Thanks,

mattclay commented on test/integration/targets/connection_sshchroot/test_connection.inventory:
```diff
@@ -0,0 +1,9 @@
+ansible_host=localhost
+ansible_connection=sshchroot
+chroot_path=/
+chroot_executable=/usr/sbin/chroot
```
The openSUSE images use `/usr/bin/chroot` instead, so you'll need to generate this file from a template and fill in the correct path to the `chroot` binary.

aarnaud replied:
Fixed I add extra vars with the `chroot` path by using `which`

mkrizek commented:
Thank you for your interest in Ansible! We are no longer accepting new plugins to ansible/ansible directly. Please re-file this pull request against https://www.github.com/ansible-collections/community.general.

If you have further questions please stop by IRC or the mailing list:
* IRC: #ansible on irc.freenode.net
* mailing list: https://groups.google.com/forum/#!forum/ansible-project
